### PR TITLE
Vectorisation

### PR DESCRIPTION
**Pull request by sv2518**
I would like to merge TJs changes to enable vectorisation of TSFC kernels.

This branch depends on https://github.com/firedrakeproject/loopy/tree/cvec.

#590 in PyOP2 (+ https://github.com/firedrakeproject/firedrake/tree/wence/fix/interpolate-max in firedrake) should be merged first, so I can drop those commits in this branch.

### CODE REVIEW

File: pyop2/compilation.py [hunk 1]
```diff
@@ -394,7 +394,7 @@ class LinuxCompiler(Compiler):
     :kwarg comm: Optional communicator to compile the code on (only
     rank 0 compiles code) (defaults to COMM_WORLD)."""
     def __init__(self, cppargs=[], ldargs=[], cpp=False, comm=None):
-        opt_flags = ['-march=native', '-O3', '-ffast-math']
+        opt_flags = ['-O3', '-ffast-math', '-fopenmp']
```
[the remark this answers is not in the document]
Comment (wence-) [reply pyop2/compilation.py]: They both just want `-fopenmp-simd`, right. And I think intel wants `-qopenmp-simd`.

**Comment by sv2518** [comment]
The firedrake test test_real_space.py::test_real_interpolate is failing with the following generated code and the answer is 4 times the answer of what we expect. It seems like there is another bug in the splitting. I think the iname should not be split at all, it is only of extent 1.
```
---------------------------------------------------------------------------
KERNEL: wrap_expression_kernel
---------------------------------------------------------------------------
ARGUMENTS:
end: ValueArg, type: np:dtype('int32')
glob0: type: np:dtype('float64'), shape: (1), dim_tags: (N0:stride:1) aspace: global
glob1: type: np:dtype('float64'), shape: (1), dim_tags: (N0:stride:1) aspace: global
start: ValueArg, type: np:dtype('int32')
---------------------------------------------------------------------------
DOMAINS:
[end, start] -> { [n_outer, n_batch] : n_batch >= start - 4n_outer and 0 <= n_batch <= 3 and n_batch < end - 4n_outer }
{ [expr_p0] : expr_p0 = 0 }
---------------------------------------------------------------------------
INAME IMPLEMENTATION TAGS:
expr_p0: None
n_batch: c_vec
n_outer: None
---------------------------------------------------------------------------
TEMPORARIES:
_zeros: type: np:dtype('float64'), shape: () scope:global
---------------------------------------------------------------------------
INSTRUCTIONS:
    for n_outer, n_batch
↱       glob0[0] = 0  {id=statement1}
├↱      ... nop  {id=expr__start}
││      for expr_p0
└└↱       glob0[expr_p0] = glob0[expr_p0] + glob1[0]  {id=expr_insn}
  │     end expr_p0
  └     ... nop  {id=statement0}
    end n_outer, n_batch
---------------------------------------------------------------------------
```

File: pyop2/compilation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sv2518) [reply pyop2/compilation.py]: Why did you turn off '-march=native', @tj-sun ?

**Comment by wence-** [comment]
Can you dump the wrapper kernel _before_ this splitting transform so we can have a look at what went wrong? Do `loopy.dump_as_python(kernel)` to get something that can be used to construct it independently.

**Comment by sv2518** [comment]
```
import loopy as lp
import numpy as np

knl = lp.make_kernel(
    [
    "[end, start] -> { [n] : start <= n < end }",
    "{ [expr_p0] : expr_p0 = 0 }",
    ],
    """
    glob0[0] = 0 {id=statement1, inames=n}
    ... nop {id=expr__start, dep=statement1, inames=n}
    glob0[expr_p0] = glob0[expr_p0] + glob1[0] {id=expr_insn, dep=expr__start:statement1, inames=n:expr_p0}
    ... nop {id=statement0, dep=expr_insn, inames=n}
    """, [
        lp.ValueArg(
            name='start', dtype=np.int32),
        lp.ValueArg(
            name='end', dtype=np.int32),
        lp.GlobalArg(
            name='glob0', dtype=np.float64,
            shape=(1,), for_atomic=False),
        lp.GlobalArg(
            name='glob1', dtype=np.float64,
            shape=(1,), for_atomic=False),
        lp.TemporaryVariable(
            name='_zeros', dtype=np.float64,
            shape=(), for_atomic=False,
            address_space=lp.AddressSpace.GLOBAL,
            read_only=True,
            initializer=np.array(0.),
            ),
        ], lang_version=(2018, 2))
```

File: pyop2/configuration.py [hunk 2]
```diff
@@ -76,6 +76,7 @@ class Configuration(dict):
     DEFAULTS = {
         "compiler": ("PYOP2_BACKEND_COMPILER", str, "gcc"),
         "simd_width": ("PYOP2_SIMD_WIDTH", int, 4),
+        "vectorization_strategy":("PYOP2_VECT_STRATEGY", str, "ve"),
```
[the remark this answers is not in the document]
Comment (sv2518) [reply pyop2/configuration.py]: How do I get the simd width from the environment? I only found information on how you get the processor type. On a Mac I can do `sysctl -a | grep machdep.cpu.features` to see which extensions are available, but the command is different between platforms.

[the remark this answers is not in the document]
Comment (wence-) [reply pyop2/configuration.py]: You can inspect the output of the `cpuid` instruction. e.g. https://github.com/flababah/cpuid.py

Or you can use [`py-cpuinfo`](https://github.com/workhorsy/py-cpuinfo) which just wraps all of the OS-specific query tools. The former is more low-level and lightweight, it just carves out some memory and splats executable code into it, the latter is probably a little safer?

File: pyop2/sequential.py [hunk 3]
```diff
@@ -57,8 +59,47 @@
...
+    kernel = loopy.split_iname(kernel, iname, batch_size, slabs=(0, 1), inner_tag="c_vec", inner_iname=inner_iname)
+
+    alignment = configuration["alignment"]
+    tmps = dict((name, tv.copy(alignment=alignment)) for name, tv in kernel.temporary_variables.items())
+    kernel = kernel.copy(temporary_variables=tmps)
+
+    wrapper = wrapper.with_root_kernel(kernel)
+
+    vec_types = [("double", 8), ("int", 4)]
+    preamble = ["typedef {0} {0}{1} __attribute__ ((vector_size ({2})));".format(t, batch_size, batch_size*b) for t, b in vec_types]
```
[the remark this answers is not in the document]
Comment (sv2518) [reply pyop2/sequential.py]: I think we can attach a preamble_generator to the `CVecASTBuilder`. The vector target would need to know the simd width, however. Would it be reasonable to pass the simd width into the initialisation of the `CVecTarget`?

**Comment by wence-** [comment]
The loop splitting code doesn't seem to privatise the reduction variables properly for global reductions.

We start with (I've assumed start = 0, end mod 4 = 0 and end > start to remove some conditionals):

```c
void loopy_kernel(int32_t const start, int32_t const end,
                  double *__restrict__ glob0,
                  double const *__restrict__ glob1) {
  for (int32_t n = 0; n <= -1 + end; ++n) {
    glob0[0] = 0.0;
    {
      int32_t const expr_p0 = 0;

      glob0[0] = glob0[0] + glob1[0];
    }
  }
}
```

So far, so good.

If I `lp.split_iname(knl, "n", 4, inner_tag="c_vec")`, then after code generation I get

```c
void loopy_kernel(int32_t const start, int32_t const end,
                  double *__restrict__ glob0,
                  double const *__restrict__ glob1) {
  for (int32_t n_outer = 0; n_outer <= (-4 + end) / 4; ++n_outer) {
    for (int32_t n_inner = 0; n_inner <= 3; ++n_inner) {
      glob0[0] = 0.0;
    }
    {
      int32_t const expr_p0 = 0;

      for (int32_t n_inner = 0; n_inner <= 3; ++n_inner)
        glob0[0] = glob0[0] + glob1[0];
    }
  }
}
```

Notice how this is wrong, because we zero `glob0[0]` four times and then sum into it four times, rather than zeroing on every loop iteration at the same time as summing. This occurs, I think, because loopy doesn't know that the zeroing of `glob0` is really in the same loop nest as the summation (because the former is created by the pyop2 wrapper gen, and the latter by tsfc, so they have different effective inames).

I wonder what one should have done.

File: pyop2/configuration.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sv2518) [reply pyop2/configuration.py]: Okay, where exactly do I need to add the install for py-cpuinfo?

[the remark this answers is not in the document]
Comment (wence-) [reply pyop2/configuration.py]: Add it to `requirements-ext.txt` in this repo, then it will be automatically pip installed I believe.

**Comment by sv2518** [comment]
I don't know. Which instruction belongs to which loop is usually defined with within_inames. Maybe we missed to specify that somewhere. But since the the instruction is in the loop body, I assume this is not it. Maybe the fact that it's wrapped in a `Block` is the problem.

**Comment by wence-** [comment]
> I don't know. Which instruction belongs to which loop is usually defined with within_inames. Maybe we missed to specify that somewhere. But since the the instruction is in the loop body, I assume this is not it. Maybe the fact that it's wrapped in a Block is the problem.

The problem here is that the `expr_p0` iname comes from tsfc, whereas the `glob0[0] = 0` assignment from the wrapper. If they had the same iname, it would be OK, but we can't make that happen. I reported a loopy issue here: https://gitlab.tiker.net/inducer/loopy/-/issues/217

**Comment by wence-** [comment]
A workaround is to add yet another conditional to the case in which we switch vectorisation off again, see https://github.com/OP2/PyOP2/pull/589/files#diff-c87324f585685c592b11a1e78337d775R183

**Comment by sv2518** [comment]
Do you expect this only to occur for writing into globals? I added the following conditional and was hoping the other failing test passes as well, but no.
```
has_wg = any(isinstance(arg.data, Global) and arg.access == WRITE for arg in self._args)
if isinstance(self._kernel.code, loopy.LoopKernel) and not (has_matrix or has_rw or has_wg):
```

**Comment by wence-** [comment]
No clue, but I'm not sure that it should only trigger then the arg is `WRITE`, I suspect it should trigger when the arg is not `READ`

**Comment by wence-** [comment]
All of these workarounds are not very satisfactory solutions: it feels like the implementation of the cvec and omp_simd backends in loopy are not very complete. But I guess one should get as much code merged as possible and then fix that.

**Comment by sv2518** [comment]
Yes to both comments.

The line smoother test is still failing. It's only the one for the periodic hierarchy. The solver diverges in the end. I have no idea how to debug this.

**Comment by wence-** [comment]
> The line smoother test is still failing. It's only the one for the periodic hierarchy. The solver diverges in the end. I have no idea how to debug this.

1. Check the assembly of the operator is correct. 
2. Check the residual
3. Can you get failure on a mesh with a single column? (It will take a different number of iterations from what the test asserts)

File: pyop2/compilation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tj-sun) [reply pyop2/compilation.py]: Hmm.. good question. I can't recall why, it's probably a mistake.

I don't remember I tried `-fopenmp-simd`, I seem to remember I need to do`-fopenmp=something` for clang, but I could be wrong.

[the remark this answers is not in the document]
Comment (wence-) [reply pyop2/compilation.py]: https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fopenmp-simd

**Comment by wence-** [comment]
> A workaround is to add yet another conditional to the case in which we switch vectorisation off again, see

Here's a proposed better fix. The loopy folks suggest privatizing (via a temporary variable) the reduction (see discussion here: https://gitlab.tiker.net/inducer/loopy/-/issues/217).

We can do this by producing code with temporaries at codegen time. I pushed 0677a24 which does this, and hopefully fixes the problem (and doesn't introduce other ones).

**Comment by wence-** [comment]
OK, that was a bad fix, will revert for now.

File: pyop2/sequential.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sv2518) [reply pyop2/sequential.py]: I fixed this here within loopy here https://github.com/firedrakeproject/loopy/pull/6/commits/8a3a220de526d63c27344410693c86d31abec705.

**Comment by sv2518** [comment]
What went wrong with your fix, @wence- ?

**Comment by wence-** [comment]
Oh, it revealed firedrakeproject/firedrake#1768.

File: pyop2/sequential.py [hunk 4]
```diff
@@ -57,8 +59,47 @@
...
+        return wrapper
+
+    # create constant zero vectors
+    wrapper = wrapper.copy(target=loopy.CVecTarget())
+    kernel = wrapper.root_kernel
+    zeros = loopy.TemporaryVariable("_zeros", shape=loopy.auto, dtype=numpy.float64, read_only=True,
+                                    initializer=numpy.array(0.0, dtype=numpy.float64),
+                                    address_space=loopy.AddressSpace.GLOBAL, zero_size=batch_size)
+    tmps = kernel.temporary_variables.copy()
+    tmps["_zeros"] = zeros
```
[the remark this answers is not in the document]
Comment (sv2518) [reply pyop2/sequential.py]: The code generation for CVecTargets was in dependence of the naming of this variable here. I changed that by moving the temporary variable generation into the code generation for CVecTargets in loopy.

**Comment by sv2518** [comment]
The kernel, that causes the failure in tests/regression/test_line_smoother_periodic.py::test_line_smoother_periodic is attached in vectorised and unvectorised form. The bug is in the value of `n_outer` in the initial slab.

[Archive.zip](https://github.com/OP2/PyOP2/files/4880148/Archive.zip)

**Comment by wence-** [comment]
Can you show what values the `start` and `end` parameters have that result in bad values for `n_outer` in the initial slab?

Translating that loop structure to python I get:

```
def single(start, end):
    for i in range(start, end):
        print(i)


def fdivpos(a, b):
    if a < 0:
        a = a - (b - 1)
    return a // b


def unroll(start, end):
    n_outer = start - ((3 + 3*start) // 4)
    print("initial")
    if end - start - 1 >= 0:
        for b in range(start - 4*n_outer, 4 if end - 4*(n_outer+1) >= 0 else end - 4*n_outer):
            print(4*n_outer + b)
    print("bulk")
    for n in range(1 + start - ((3 + 3*start)//4), (3+end)//4 - 1):
        for b in range(4):
            print(4*n + b)

    print("final")
    n_outer = fdivpos(end - 1, 4)
    if 4*n_outer - start - 1 >= 0:
        for b in range(0, end - 4*n_outer):
            print(4*n_outer + b)
```

It's not obvious that this gives wrong iteration counts.

**Comment by sv2518** [comment]
I think, the reason one cannot see the problem in the iteration count in your script is that the slab `n_outer` in the C script is  differently rounded. To reproduce the problem you would need to exchange `n_outer = start - ((3 + 3*start) // 4)` with `math.floor(start - ((3 + 3*start) / 4))`. In that case the iterations over the initial slab are missing.
Start and end sets in the test are: (0,18), (18,18), (0,72), (72,72).

**Comment by sv2518** [comment]
So I think in the C kernel we want `int const n_outer = start - loopy_floor_div_pos_b_int32((3 + 3 * start), 4);`

**Comment by wence-** [comment]
> I think, the reason one cannot see the problem in the iteration count in your script is that the slab n_outer in the C kernel is differently rounded. To reproduce the problem you would need to exchange n_outer = start - ((3 + 3*start) // 4) with math.floor(start - ((3 + 3*start) / 4)). In that case the iterations over the initial slab are missing.
Start and end sets in the test are: (0,18), (18,18), (0,72), (72,72).

I'm pretty sure python and C ints have the same rounding behaviour.

I did a bit more digging, it turns out this is not a bug in the loopy codegen, but rather in _values_ that we're sending in to the kernel.

You can reproduce by comparing the output with vectorisation off and on of this code:
```
from firedrake import *


def periodise(m):
    coord_fs = VectorFunctionSpace(m, "DG", 1, dim=2)
    old_coordinates = m.coordinates
    new_coordinates = Function(coord_fs)
    domain = "{[i, j]: 0 <= i < old_coords.dofs and 0 <= j < new_coords.dofs}"
    instructions = """
    <float64> Y = 0
    <float64> pi = 3.141592653589793
    for i
        Y = Y + old_coords[i, 1]
    end
    for j
        new_coords[j, 0] = atan2(old_coords[j, 1], old_coords[j, 0]) / (pi* 2)
        new_coords[j, 0] = new_coords[j, 0] + (1 if new_coords[j, 0] < 0 else 0)
        new_coords[j, 0] = 1 if (new_coords[j, 0] == 0 and Y < 0) else new_coords[j, 0]
        new_coords[j, 0] = new_coords[j, 0] * Lx[0]
        new_coords[j, 1] = old_coords[j, 2] * Ly[0]
    end
    """
    cLx = Constant(1)
    cLy = Constant(1)
    par_loop((domain, instructions), dx,
             {"new_coords": (new_coordinates, WRITE),
              "old_coords": (old_coordinates, READ),
              "Lx": (cLx, READ),
              "Ly": (cLy, READ)},
             is_loopy_kernel=True)
    return Mesh(new_coordinates, reorder=False)

H = 1
N = 3

mesh = periodise(CylinderMesh(N, N, 1.0, H, reorder=False))

V = FunctionSpace(mesh, "P", 1)
x, _ = SpatialCoordinate(mesh)
xc = interpolate(x, V)

print(xc.dat.data_ro)

# "working"
=> [0.66666667 1.         0.66666667 0.33333333 0.33333333 0.
 0.         0.66666667 0.33333333 0.66666667 0.33333333 0.        ]

# "broken"
=> [0.66666667 1.         0.66666667 0.33333333 0.33333333 1.
 0.         0.66666667 0.33333333 0.66666667 0.33333333 0.        ]
```

What's happening here is that we're interpolating a DG field into a CG space, and the values we get are dependent on the order in which we execute things. Now, PyOP2 has never guaranteed that it won't change the iteration order, so arguably this code has been relying on unspecified behaviour.

Actually fixing this properly needs a big rethink of the way the multigrid grid transfer is done, however, I can get the test to pass at least, by applying this patch in loopy:

```patch
diff --git a/loopy/codegen/loop.py b/loopy/codegen/loop.py
index c71554a4..43bff2c3 100644
--- a/loopy/codegen/loop.py
+++ b/loopy/codegen/loop.py
@@ -105,9 +105,9 @@ def get_slab_decomposition(kernel, iname):
         if upper_bulk_bound is not None:
             bulk_slab = bulk_slab.add_constraint(upper_bulk_bound)
 
-        slabs.append(("bulk", bulk_slab))
         if lower_slab:
             slabs.append(lower_slab)
+        slabs.append(("bulk", bulk_slab))
         if upper_slab:
             slabs.append(upper_slab)
```

**Comment by sv2518** [comment]
If I execute that test I get a different output

```
"broken"
[0.         0.33333333 0.         0.33333333 0.66666667 0.66666667
 1.         0.33333333 0.66666667 1.         0.33333333 0.66666667]
"working, including your temporary fix":
[1.         0.33333333 1.         0.33333333 0.66666667 0.66666667
 1.         0.33333333 0.66666667 1.         0.33333333 0.66666667]
```

**Comment by sv2518** [comment]
The last failing firedrake test is tests/extrusion/test_real_tensorproduct.py now.

**Comment by sv2518** [comment]
#590 in PyOP2 (+ https://github.com/firedrakeproject/firedrake/tree/wence/fix/interpolate-max in firedrake) should be merged first, so I can drop those commits in this branch.

**Comment by wence-** [comment]
> #590 in PyOP2 (+ https://github.com/firedrakeproject/firedrake/tree/wence/fix/interpolate-max in firedrake) should be merged first, so I can drop those commits in this branch.

Done.

**Comment by sv2518** [comment]
I have py-cpuinfo as required dependency. I added this in firedrake https://github.com/firedrakeproject/firedrake/compare/trigger-tests-for-vec and also in PyOP2 in this commit https://github.com/OP2/PyOP2/pull/589/commits/1cf7698593e77db83859f229183540ea931227c0. Do I need both or is it enough in PyOP2?

**Comment by sv2518** [comment]
Superseded by  #654

**Pull request closed without merging**